numpy/core/src/multiarray/nditer_pywrap.c

```diff
@@ -1173,10 +1174,29 @@ NpyIter_NestedIters(PyObject *NPY_UNUSED(self),
     return NULL;
 }
 
+
 static void
 npyiter_dealloc(NewNpyArrayIterObject *self)
 {
     if (self->iter) {
+        if (npyiter_has_writeback(self->iter)) {
```

Comment:
Ideally this would move to `tp_finalize`, once we drop 2.7. Then you could also just pass `npiter.__del__` in as the argument to WriteUnraiseable.

We'd use the `PyObject_CallFinalizerFromDealloc` function, which is undocumented.